BUG: test fail when run a second time

### Describe the issue:

When running numpy.test() - all tests pass as expected on the first run. If you attempt to run the command a second time in the same session, 4 tests fail (presumably because of pre-existing data?). Quit python and run tests for the first time, everything again works the first time and fails again on the second attempt.
Test was done with local build as well as pip installed wheel with the same results.

### Reproduce the code example:

```python
import numpy
numpy.test()
numpy.test()
```


### Error message:

```shell
FAILED _core/tests/test_dtype.py::TestPickling::test_pickle_dtype_class[dtype[rational]] - _pickle.PicklingError: Can't pickle <function _DType_reconstruct at 0x7fffb67dd4e0>: it's not the same object as numpy._core._DType_reconstruct
FAILED f2py/tests/test_f2py2e.py::test_gen_pyf_no_overwrite - ValueError: I/O operation on closed file.
FAILED lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[1-data2] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) matching the regex were emitted.
FAILED lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[0-data5] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) matching the regex were emitted.
```


### Python and NumPy Versions:

python 3.12.3
numpy 2.0.0

### Runtime Environment:

[{'numpy_version': '2.0.0',
  'python': '3.12.3 (main, Apr 29 2024, 13:56:22) [GCC 12.3.0]',
  'uname': uname_result(system='Linux', node='c002', release='5.14.21-150400.24.28-default', version='#1 SMP PREEMPT_DYNAMIC Mon Oct 10 15:21:12 UTC 2022 (f82da2c)', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE',
                                   'SSE2',
                                   'SSE3',
                                   'SSSE3',
                                   'SSE41',
                                   'POPCNT',
                                   'SSE42',
                                   'AVX',
                                   'F16C',
                                   'FMA3',
                                   'AVX2'],
                      'found': [],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL',
                                    'AVX512_SPR']}},
 {'architecture': 'ZEN',
  'filepath': '/opt/openblas/0.3.24/lib/libopenblas_zenp-r0.3.24.so',
  'internal_api': 'openblas',
  'num_threads': 8,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.24'}]

### Context for the issue:

low priority I suspect? Who runs tests twice in the same session? But perhaps tests could be improved to handle this scenario (if it's just a test case issue)?

I’m struggling to debug this one. With `numpy==2.0.0` I can replicate the issue. I’ve done a local installation with `2.1.0.dev0+git20240713.c1d1ff4` but when I run `numpy.test()` I get errors like with `disutils`

```
______________________ ERROR collecting numpy/distutils/tests/test_system_info.py ______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_system_info.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
```

I wanted debugging in IDE support so was following instructions here https://numpy.org/doc/stable/building/index.html#building-from-source-for-numpy-development

```sh
git clean -fdx
uv venv
source .venv/bin/activate
uv pip install -r requirements/all_requirements.txt
uv pip install -e . --no-build-isolation
python -c "import numpy; numpy.__version; numpy.test()"
```

which leads to
```
NumPy version 2.1.0.dev0+git20240714.2093a6d
NumPy CPU features:  NEON NEON_FP16 NEON_VFPV4 ASIMD ASIMDHP* ASIMDFHM?

================================================ ERRORS ================================================
_______________________ ERROR collecting numpy/distutils/tests/test_build_ext.py _______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_build_ext.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_____________________ ERROR collecting numpy/distutils/tests/test_ccompiler_opt.py _____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_ccompiler_opt.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
__________________ ERROR collecting numpy/distutils/tests/test_ccompiler_opt_conf.py ___________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_ccompiler_opt_conf.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_____________________ ERROR collecting numpy/distutils/tests/test_exec_command.py ______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_exec_command.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_______________________ ERROR collecting numpy/distutils/tests/test_fcompiler.py _______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_fcompiler.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_____________________ ERROR collecting numpy/distutils/tests/test_fcompiler_gnu.py _____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_fcompiler_gnu.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
____________________ ERROR collecting numpy/distutils/tests/test_fcompiler_intel.py ____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_fcompiler_intel.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
___________________ ERROR collecting numpy/distutils/tests/test_fcompiler_nagfor.py ____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_fcompiler_nagfor.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_____________________ ERROR collecting numpy/distutils/tests/test_from_template.py _____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_from_template.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
__________________________ ERROR collecting numpy/distutils/tests/test_log.py __________________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_log.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
___________________ ERROR collecting numpy/distutils/tests/test_mingw32ccompiler.py ____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_mingw32ccompiler.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
_______________________ ERROR collecting numpy/distutils/tests/test_misc_util.py _______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_misc_util.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
____________________ ERROR collecting numpy/distutils/tests/test_npy_pkg_config.py _____________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_npy_pkg_config.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
______________________ ERROR collecting numpy/distutils/tests/test_shell_utils.py ______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_shell_utils.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
______________________ ERROR collecting numpy/distutils/tests/test_system_info.py ______________________
ImportError while importing test module '/Users/paddy/repos/numpy/numpy/distutils/tests/test_system_info.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'numpy.distutils'
======================================= short test summary info ========================================
ERROR numpy/distutils/tests/test_build_ext.py
ERROR numpy/distutils/tests/test_ccompiler_opt.py
ERROR numpy/distutils/tests/test_ccompiler_opt_conf.py
ERROR numpy/distutils/tests/test_exec_command.py
ERROR numpy/distutils/tests/test_fcompiler.py
ERROR numpy/distutils/tests/test_fcompiler_gnu.py
ERROR numpy/distutils/tests/test_fcompiler_intel.py
ERROR numpy/distutils/tests/test_fcompiler_nagfor.py
ERROR numpy/distutils/tests/test_from_template.py
ERROR numpy/distutils/tests/test_log.py
ERROR numpy/distutils/tests/test_mingw32ccompiler.py
ERROR numpy/distutils/tests/test_misc_util.py
ERROR numpy/distutils/tests/test_npy_pkg_config.py
ERROR numpy/distutils/tests/test_shell_utils.py
ERROR numpy/distutils/tests/test_system_info.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 15 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
34 skipped, 2624 deselected, 15 errors in 20.61s

numpy on  main [$] is 📦 v2.1.0.dev0 via 🐍 v3.12.4 (numpy) took 23s
```

My apologies - no - this is still happening with 2.2.0:
First run - all tests pass
second run:

`FAILED _core/tests/test_deprecations.py::TestAddNewdocUFunc::test_deprecated - ValueError: Cannot change docstring of ufunc with non-NULL docstring
FAILED _core/tests/test_dtype.py::TestPickling::test_pickle_dtype_class[dtype[rational]] - _pickle.PicklingError: Can't pickle <function _DType_reconstruct at 0x7fff031b63e0>: it's not the same object as numpy._core._DType_reconstruct
FAILED f2py/tests/test_crackfortran.py::TestPublicPrivate::test_defaultPrivate - KeyError: 'a'
FAILED f2py/tests/test_crackfortran.py::TestPublicPrivate::test_defaultPublic - KeyError: 'a'
FAILED f2py/tests/test_crackfortran.py::TestPublicPrivate::test_access_type - KeyError: 'a'
FAILED f2py/tests/test_crackfortran.py::TestModuleProcedure::test_moduleOperators - AssertionError: assert ('body' in {'args': [], 'block': 'python module', 'body': [{'args': [], 'block': 'interface', 'body': [{'args': [], 'block': 'mod...dy': [{...}, {...}, {...}, {...}, {...}, {...}, ...], 'entry': {}, ...}], 'externals': [], ...}], 'externals': [], ...} and 1 == 9)
FAILED f2py/tests/test_crackfortran.py::TestModuleProcedure::test_notPublicPrivate - KeyError: 'a'
FAILED f2py/tests/test_crackfortran.py::TestModuleDeclaration::test_dependencies - KeyError: 'abar'
FAILED f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[ascii] - AssertionError: assert 'blah' == 'foo'
FAILED f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-8] - AssertionError: assert 'blah' == 'foo'
FAILED f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-16] - AssertionError: assert 'blah' == 'foo'
FAILED f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-32] - AssertionError: assert 'blah' == 'foo'
FAILED f2py/tests/test_data.py::TestData::test_crackedlines - KeyError: 'x'
FAILED f2py/tests/test_data.py::TestDataF77::test_crackedlines - KeyError: 'mydata'
FAILED lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[1-data2] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) matching the regex were emitted.
FAILED lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[0-data5] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.`

Without even being in a development environment, I get the following after the second `numpy.test()`

```
==================================================================== short test summary info ====================================================================
FAILED numpy/_core/tests/test_deprecations.py::TestAddNewdocUFunc::test_deprecated - ValueError: Cannot change docstring of ufunc with non-NULL docstring
FAILED numpy/_core/tests/test_dtype.py::TestPickling::test_pickle_dtype_class[dtype[rational]] - _pickle.PicklingError: Can't pickle <function _DType_reconstruct at 0x71d3a63823e0>: it's not the same object as numpy._core._DType_reconstruct
FAILED numpy/f2py/tests/test_crackfortran.py::TestPublicPrivate::test_defaultPrivate - KeyError: 'a'
FAILED numpy/f2py/tests/test_crackfortran.py::TestPublicPrivate::test_defaultPublic - KeyError: 'a'
FAILED numpy/f2py/tests/test_crackfortran.py::TestPublicPrivate::test_access_type - KeyError: 'a'
FAILED numpy/f2py/tests/test_crackfortran.py::TestModuleProcedure::test_moduleOperators - AssertionError: assert ('body' in {'args': [], 'block': 'python module', 'body': [{'args': [], 'block': 'interface', 'body': [{'args': [], 'block': 'mod...d...
FAILED numpy/f2py/tests/test_crackfortran.py::TestModuleProcedure::test_notPublicPrivate - KeyError: 'a'
FAILED numpy/f2py/tests/test_crackfortran.py::TestModuleDeclaration::test_dependencies - KeyError: 'abar'
FAILED numpy/f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[ascii] - AssertionError: assert 'blah' == 'foo'
FAILED numpy/f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-8] - AssertionError: assert 'blah' == 'foo'
FAILED numpy/f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-16] - AssertionError: assert 'blah' == 'foo'
FAILED numpy/f2py/tests/test_crackfortran.py::TestFortranReader::test_input_encoding[utf-32] - AssertionError: assert 'blah' == 'foo'
FAILED numpy/f2py/tests/test_data.py::TestData::test_crackedlines - KeyError: 'x'
FAILED numpy/f2py/tests/test_data.py::TestDataF77::test_crackedlines - KeyError: 'mydata'
FAILED numpy/lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[1-data2] - AssertionError: 
FAILED numpy/lib/tests/test_io.py::TestLoadTxt::test_max_rows_empty_lines[0-data5] - AssertionError: 
16 failed, 45296 passed, 1750 skipped, 1485 deselected, 32 xfailed, 5 xpassed in 162.47s (0:02:42)
```

I found 1 more with `numpy.test(label = "full")`
```
FAILED /numpy/f2py/tests/test_docs.py::TestDocAdvanced::test_ftype
```
I've isolated the cause of all 17 failing tests that occur when running `numpy.test(label = "full")` twice in the same Python session.

1. `_core/tests` - Immutable in `TestAddNewdocUfunc`
    * Problem: `TestAddNewdocUfunc` mutates the C-level docstring pointer for `add_triplet`, which is immutable once set. Resolving this fixes the pickling error. 
    * Solution: Skip the test if the pointer has previously been set. 
    * I tried using `@pytest.mark.forked` but this feature doesn't work on windows and requires extra overhead.  
    * A more robust solution might involve running this test in a subprocess. However, skipping on repeat seems sufficient.
2. `f2py/tests` - Stale global `crackfortran` state
    * All issues stem from the mutable global `crackfortran.f77modulename`.
    * This is set in `f2py2e.callcrackfortran` but never reset, causing a stale state. In particular, running `test.f2py2e` followed by `test.crack_fortran` shows the issue. 
    * Solution: Add `crackfortran.f77modulename = ''` at the end of `f2py2e.callcrackfortran`.     
    * This changes the behavior of `f2py2e.callcrackfortran`, so may need a discussion. If accepted, I'll write a unit test.
    * Alternative (less desirable): patch over the root cause using `setup_module()` + `reset_global_f2py_vars()`.  This hides the root cause. 
    * Note: In addition to `groupcache` mentioned above, several global variables are not reset with `reset_global_f2py_vars()`. This might warrant a different issue. The global state is constantly in flux in the entire f2py module. 
3. `lib/tests` - Parametrize reuses `StringIO`
    * Problem: The parametrize option for `TestLoadTxt` contains two `StringIO` objects. Their state persists on second run.  
    * Solution: Wrap these two lines in `lambda` to ensure a fresh instance on each run. 
4. `f2py/tests/test_docs` - Stale state of `ftype.data`
    * Problem: The first assertion `assert_equal(ftype.data.a, 0)` checks instantiation, but the later `ftype.data.a = 3` persists across runs.
    * Solution: Insert `ftype.data.a = 0` at the end of `test_ftype`.
    * We could repeatedly test proper instantiation, but I think this is sufficient.

PR incoming.

--- a/numpy/_core/tests/test_deprecations.py
+++ b/numpy/_core/tests/test_deprecations.py
@@ -448,6 +448,13 @@ class TestDeprecatedSaveFixImports(_DeprecationTestCase):
 class TestAddNewdocUFunc(_DeprecationTestCase):
     # Deprecated in Numpy 2.2, 2024-11
     def test_deprecated(self):
+        doc = struct_ufunc.add_triplet.__doc__
+        # gh-26718
+        # This test mutates the C-level docstring pointer for add_triplet,
+        # which is permanent once set. Skip when re-running tests.
+        if doc is not None and "new docs" in doc:
+            pytest.skip("Cannot retest deprecation, otherwise ValueError: "
+                "Cannot change docstring of ufunc with non-NULL docstring")
         self.assert_deprecated(
             lambda: np._core.umath._add_newdoc_ufunc(
                 struct_ufunc.add_triplet, "new docs"

--- a/numpy/f2py/f2py2e.py
+++ b/numpy/f2py/f2py2e.py
@@ -378,6 +378,8 @@ def callcrackfortran(files, options):
             mod['gil_used'] = 'Py_MOD_GIL_USED'
         else:
             mod['gil_used'] = 'Py_MOD_GIL_NOT_USED'
+    # gh-26718 Reset global
+    crackfortran.f77modulename = ''
     return postlist
 
 

--- a/numpy/f2py/tests/test_docs.py
+++ b/numpy/f2py/tests/test_docs.py
@@ -60,5 +60,7 @@ class TestDocAdvanced(util.F2PyTest):
         ftype.data.x[1] = 45
         assert_array_equal(ftype.data.x,
                            np.array([1, 45, 3], dtype=np.float32))
+        # gh-26718 Cleanup for repeated test runs
+        ftype.data.a = 0
 
     # TODO: implement test methods for other example Fortran codes

--- a/numpy/lib/tests/test_io.py
+++ b/numpy/lib/tests/test_io.py
@@ -1276,12 +1276,16 @@ class TestLoadTxt(LoadTxtBase):
             (1, ["ignored\n", "1,2\n", "\n", "3,4\n"]),
             # "Bad" lines that do not end in newlines:
             (1, ["ignored", "1,2", "", "3,4"]),
-            (1, StringIO("ignored\n1,2\n\n3,4")),
+            (1, lambda: StringIO("ignored\n1,2\n\n3,4")),
             # Same as above, but do not skip any lines:
             (0, ["-1,0\n", "1,2\n", "\n", "3,4\n"]),
             (0, ["-1,0", "1,2", "", "3,4"]),
-            (0, StringIO("-1,0\n1,2\n\n3,4"))])
+            (0, lambda: StringIO("-1,0\n1,2\n\n3,4"))])
     def test_max_rows_empty_lines(self, skip, data):
+        # gh-26718 re-instantiate StringIO objects each time
+        if callable(data):
+            data = data()
+
         with pytest.warns(UserWarning,
                     match=f"Input line 3.*max_rows={3 - skip}"):
             res = np.loadtxt(data, dtype=int, skiprows=skip, delimiter=",",

